source/snapshot/iterator.go

```diff
@@ -0,0 +1,207 @@
+// Copyright © 2022 Meroxa, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package snapshot
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/conduitio-labs/conduit-connector-vitess/coltypes"
+	sdk "github.com/conduitio/conduit-connector-sdk"
+	"github.com/huandu/go-sqlbuilder"
+	"github.com/jmoiron/sqlx"
+	"vitess.io/vitess/go/vt/vitessdriver"
+)
+
+const (
+	// metadataKeyTable is a metadata key for table.
+	metadataKeyTable = "table"
+	// metadataKeyTable is a metadata key for action.
+	metadataKeyAction = "action"
+
+	// actionInsert is a value for the metadataKeyAction that represents insert action.
+	actionInsert = "insert"
+)
+
+// Iterator is an implementation of a Snapshot iterator for Vitess.
+type Iterator struct {
+	db   *sqlx.DB
+	rows *sqlx.Rows
+
+	table          string
+	keyColumn      string
+	orderingColumn string
+	columns        []string
+	columnTypes    map[string]string
+	batchSize      int
+	position       *Position
+}
+
+// SnaphostParams is an incoming params for the NewIterator function.
+type IteratorParams struct {
+	Address        string
+	Target         string
+	Table          string
+	KeyColumn      string
+	OrderingColumn string
+	Columns        []string
+	BatchSize      int
+	Position       sdk.Position
+}
+
+// NewIterator create a new instance of the Iterator iterator.
+func NewIterator(ctx context.Context, params IteratorParams) (*Iterator, error) {
+	db, err := vitessdriver.Open(params.Address, params.Target)
+	if err != nil {
+		return nil, fmt.Errorf("connect to vtgate: %w", err)
+	}
+
+	if err := db.PingContext(ctx); err != nil {
+		return nil, fmt.Errorf("ping vtgate: %w", err)
+	}
+
+	iterator := &Iterator{
+		db:             sqlx.NewDb(db, "vitess"),
+		table:          params.Table,
+		keyColumn:      params.KeyColumn,
+		orderingColumn: params.OrderingColumn,
+		columns:        params.Columns,
+		batchSize:      params.BatchSize,
+	}
+
+	if params.Position != nil {
+		position, err := parsePosition(params.Position)
+		if err != nil {
```

Comment:
err => er

Reply:
Sorry, didn't get you

Reply:
please rename err to er or something else, or declare var err error in the start of function.
Because in this case it shadow variable